platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/loader/CachingToolingImplementationLoader.java

```diff
@@ -15,46 +15,75 @@
  */
 package org.gradle.tooling.internal.consumer.loader;
 
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.gradle.api.NonNullApi;
 import org.gradle.initialization.BuildCancellationToken;
 import org.gradle.internal.classpath.ClassPath;
 import org.gradle.internal.concurrent.CompositeStoppable;
 import org.gradle.internal.logging.progress.ProgressLoggerFactory;
+import org.gradle.tooling.GradleConnectionException;
 import org.gradle.tooling.internal.consumer.ConnectionParameters;
 import org.gradle.tooling.internal.consumer.Distribution;
 import org.gradle.tooling.internal.consumer.connection.ConsumerConnection;
 import org.gradle.tooling.internal.protocol.InternalBuildProgressListener;
 
 import java.io.Closeable;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
 
 public class CachingToolingImplementationLoader implements ToolingImplementationLoader, Closeable {
     private final ToolingImplementationLoader loader;
-    private final Map<ClassPath, ConsumerConnection> connections = new HashMap<ClassPath, ConsumerConnection>();
+    private final Cache<ClassPath, ConsumerConnection> connections = CacheBuilder.newBuilder()
+        .maximumSize(15)
+        .expireAfterAccess(7, TimeUnit.DAYS)
+        .build();
 
```

Comment:
I'm not sure we can use Guava in the client.  Do we end up shading this?

Reply:
It's at least a [dependency](https://github.com/gradle/gradle/blob/0d68f92a33bb5ccdfe7048378e3c8bf57d45dcc2/platforms/ide/tooling-api/build.gradle.kts#L46)

Any reason this wouldn't work?
Are thinking about class loading exclusion or something like that?

Reply:
Hmm, that was added a long time ago too.

Yes, I was thinking about how the classloaders would work here and if this would introduce a conflict with consumers that also use Guava (but a different version).  It looks like we shade Guava in the tooling-api jar because I see `org/gradle/internal/impldep/com/google` classes.

Reply:
We already use Guava extensively on the Tooling API consumer side. You are right, we properly shade the dependency in the generated (fat) jar to avoid conflicts with the client's classpath. This change is safe in this regard.